firebase-sessions/src/main/kotlin/com/google/firebase/sessions/api/FirebaseSessionsDependencies.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.sessions.api
+
+import androidx.annotation.VisibleForTesting
+import java.util.Collections.synchronizedMap
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * [FirebaseSessionsDependencies] determines when a dependent SDK is installed in the app. The
+ * Sessions SDK uses this to figure out which dependencies to wait for to getting the data
+ * collection state. This is thread safe.
+ *
+ * This is important because the Sessions SDK starts up before dependent SDKs.
+ */
+object FirebaseSessionsDependencies {
+  private val dependencies = synchronizedMap(mutableMapOf<SessionSubscriber.Name, Dependency>())
+
+  /**
+   * Add a subscriber as a dependency to the Sessions SDK. Every dependency must register itself, or
+   * the Sessions SDK will never generate a session.
+   */
+  fun addDependency(subscriberName: SessionSubscriber.Name) {
+    if (dependencies.containsKey(subscriberName)) {
+      throw IllegalArgumentException("Dependency $subscriberName already added.")
+    }
+
+    // The dependency is locked until the subscriber registers itself.
+    dependencies[subscriberName] = Dependency(Mutex(locked = true))
+  }
+
+  /** Register and unlock the subscriber. This must be called before [getSubscribers] can return. */
+  internal fun register(subscriber: SessionSubscriber) {
+    val subscriberName = subscriber.sessionSubscriberName
+    val dependency = getDependency(subscriberName)
+
+    dependency.subscriber?.run {
+      throw IllegalArgumentException("Subscriber $subscriberName already registered.")
+    }
+    dependency.subscriber = subscriber
+
+    // Unlock to show the subscriber has been registered, it is possible to get it now.
+    dependency.mutex.unlock()
+  }
+
+  /** Gets the subscribers safely, blocks until all the subscribers are registered. */
+  internal suspend fun getSubscribers(): Map<SessionSubscriber.Name, SessionSubscriber> {
```

Comment:
Why `suspend` - Is this to ensure that all the subscribers who are in the process of registering are complete before returning? If so, I would rather not worry about this too much and actually return the current state rather than wait for all the register/dependencies are done with the subscription?

Reply:
Do you want to call it `getRegisteredSubscribers`?

Reply:
This needs to be a suspend function because `mutex.withLock` is.

It explicitly does wait until all dependencies are registered to keep the logic in the call site simpler. We will be able to just say `val subscribers = FirebaseSessionsDependencies.getSubscribers()` and then on the next line know that `subscribers` is populated with all the dependent subscribers, they are registered, and know if they have data collection enabled.

If a subscriber enables data collection later, then sessions after that will count as data collection enabled.

Reply:
That makes sense. Sounds good to have it as a `suspend` method. Since we are calling to get all the successfully registered subscribers, do you want to call it `getRegisteredSubscribers`? This give a good comprehension of what this method does.

Reply:
Done.